server/src/main/java/org/elasticsearch/snapshots/SnapshotsService.java

```diff
@@ -824,8 +829,15 @@ public void applyClusterState(ClusterChangedEvent event) {
                 final boolean newMaster = event.previousState().nodes().isLocalNodeElectedMaster() == false;
                 processExternalChanges(
                     newMaster || removedNodesCleanupNeeded(snapshotsInProgress, event.nodesDelta().removedNodes()),
-                    event.routingTableChanged() && waitingShardsStartedOrUnassigned(snapshotsInProgress, event)
+                    snapshotsInProgress.nodeIdsForRemovalChanged(SnapshotsInProgress.get(event.previousState()))
+                        || (event.routingTableChanged() && waitingShardsStartedOrUnassigned(snapshotsInProgress, event))
```

Comment:
Nit: Can we renamed this parameter from `startShards` to `changedShards` to be more accurate? I think this was already the case, i.e. shards status changed instead of just starting new shards, before this change. Node removal just makes it even more obvious.